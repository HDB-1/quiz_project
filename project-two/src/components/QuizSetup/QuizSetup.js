import React, { Component } from 'react'

class QuizSetup extends Component {
    render() {
        return (
            <div>
                <select className="difficultyDropDown">
                    <option value="easy">Easy</option>
                    <option value="medium">Medium</option>
                    <option value="hard">Hard</option>
                </select>
                <select className="categoryDropDown">
                    <option value="9">General Knowledge</option>
                    <option value="21">Sports</option>
                    <option value="17">Science & Nature</option>
                    <option value="23">History</option>
                    <option value="28">Vehicles</option>
                </select>
                <select className = 'numberOfQuestionsDropdown'>
                    <option value="5">5</option>
                    <option value="10">10</option>
                    <option value="15">15</option>

                </select>
                <button onClick={[Function]}>This is a button</button>
            </div>
        )
    }
}

export default QuizSetup

