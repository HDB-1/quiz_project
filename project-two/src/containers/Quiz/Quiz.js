import React, { Component } from "react";
import Info from "../../components/Info/Info";
import Navigation from "../../components/Navigation/Navigation";
import Question from "../../containers/Question/Question";
import { Route, Switch, Link } from "react-router-dom";
import Results from "../../containers/Results/Results";

class Quiz extends Component {
  
  state = {
    currentQuestionIndex: 0,
    questions: [],
    correctAnswers: [],
    userAnswers: [],
    currentPlayerIndex: 0 
  };
  componentDidMount() {
    for(let i = 0; i < this.props.quizInfo.numOfPlayers; i++){
      this.state.userAnswers.push([])
    }
    this.APIRequest(this.props.quizInfo);
  }

  submitQuestion = answerSelected => {
    let answerFound = false;

    // make an answer
    let newAnswer = {
      content: answerSelected,
      index: this.state.currentQuestionIndex
    };
    let newAnswerArray = this.state.userAnswers[this.state.currentPlayerIndex]; // new answers array
    //check if index of new answer is in answer array
    //if not add it to the array
    //if so update array value
    if (newAnswerArray.length > 0) {
      // if there are any answers
      newAnswerArray.map(answer => {
        if (answer.index === newAnswer.index) {
          answer.content = newAnswer.content;
          answerFound = true;
        }
        return answer;
      });
    }
    if (!answerFound) {
      newAnswerArray.push(newAnswer);
    }
    let allUserAnswers = this.state.userAnswers;

    allUserAnswers[this.state.currentPlayerIndex] = newAnswerArray
    this.setState({ userAnswers: allUserAnswers });
    if(this.state.currentPlayerIndex === Number(this.props.quizInfo.numOfPlayers) - 1){
      this.nextQuestion();
      this.setState({currentPlayerIndex : 0})
    }
    else this.setState({currentPlayerIndex: this.state.currentPlayerIndex +1})
  };

  nextQuestion = () => {
    if (this.state.currentQuestionIndex < this.state.questions.length - 1) {
      this.setState({
        currentQuestionIndex: this.state.currentQuestionIndex + 1
      });
    }
  };
  previousQuestion = () => {
    if (this.state.currentQuestionIndex > 0) {
      this.setState({
        currentQuestionIndex: this.state.currentQuestionIndex - 1
      });
    }
  };

  APIRequest = quizInfo => {
    const baseUrl = "https://opentdb.com/api.php?type=multiple&";
    let url =
      baseUrl +
      `category=${quizInfo.category}&` +
      `amount=${quizInfo.numOfQuestions}&` +
      `difficulty=${quizInfo.difficulty}&` + 
      'encode=base64'
      ;
    // console.log(url);
    // console.log(fetch(url));
    fetch(url)
      .then(res => res.json())
      .then(json => {
        let correctAnswers = json.results.map(question => {
          return question.correct_answer;
        });
        this.setState({
          questions: json.results,
          correctAnswers: correctAnswers
        });
      })
      .catch(error => {
        //handle error
      });
  };

  checkIfUsersHaveAnsweredAllQuestions = () => {
    let counter = 0;
    for(let answerArray of this.state.userAnswers){
      if(answerArray.length === this.state.correctAnswers.length){
        counter ++;
      }
    }
    if(counter === Number(this.props.quizInfo.numOfPlayers)) {
      return true;
    }
    else return false;
  }

  render() {
    return (
      <Switch>
        <Route exact path="/quiz">
          <div className="quizPage">
            {this.state.questions.length > 0 ? (
              <div>
                <Info
                  title={this.state.questions[0].category}
                  currentUser={this.state.currentPlayerIndex}
                  question={{
                    current: this.state.currentQuestionIndex,
                    total: this.props.quizInfo.numOfQuestions
                  }}

                />
                <Question
                  skip={this.skipQuestion}
                  submit={this.submitQuestion}
                  questionInfo={
                    this.state.questions[this.state.currentQuestionIndex]
                  }
                />
                <Navigation
                  next={this.nextQuestion}
                  previous={this.previousQuestion}
                />
              </div>
            ) : (
              <p>Loading...</p>
            )}
            {this.checkIfUsersHaveAnsweredAllQuestions()  && <Link to="/quiz/results">
              You have answered every question! Click here to go to results
              </Link>}
          </div>
        </Route>
        <Route path="/quiz/results">
          <Results userAnswers={this.state.userAnswers} correctAnswers={this.state.correctAnswers}/>
        </Route>
      </Switch>
    );
  }
}

export default Quiz;
