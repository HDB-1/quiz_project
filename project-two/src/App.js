import React from "react";
import "./App.css";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import Quiz from "./containers/Quiz/Quiz";
import QuizSetup from "./components/QuizSetup/QuizSetup";
import NavBar from "./components/NavBar/NavBar";

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quizSetup: { difficulty: "easy", category: "9", numOfQuestions: "5" }
    };
    // Setting default quiz values. Eventually this should be built by a componentDidMount function in quizSetup.js
  }

  setQuizState = (quizSetupCharacteristic, characteristicValue) => {
    console.log("calling setquiz state with value" + characteristicValue);
    let quizSetupCopy = this.state.quizSetup;
    let characteristicKey = quizSetupCharacteristic;
    quizSetupCopy[characteristicKey] = characteristicValue;
    console.log(quizSetupCopy);
    this.setState({ quizSetup: quizSetupCopy });
  };

  render() {
    return (
      <Router className="App">
        <NavBar />
        <Switch>
          <Route exact path="/">
            <QuizSetup handleChange={this.setQuizState} />
          </Route>
          <Route path="/quiz">
            <Quiz quizInfo={this.state.quizSetup} />
          </Route>
        </Switch>
      </Router>
    );
  }
}

export default App;
