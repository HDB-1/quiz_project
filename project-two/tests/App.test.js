import React from "react";
import App from "../src/App";
import { shallow, mount } from "enzyme";
import { MemoryRouter, Link } from "react-router-dom";
import Quiz from "../src/containers/Quiz/Quiz";
import QuizSetup from "../src/components/QuizSetup/QuizSetup";
import NavBar from "../src/components/NavBar/NavBar";
import waitUntil from "async-wait-until"; // For async testing of setState functions.

describe("Shallow App", () => {
  let wrapper;

  beforeEach(() => (wrapper = shallow(<App />))); // this automatically tests for rendering without crashing.
  it("Should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("Should render a <Quiz /> component", () => {
    expect(wrapper.containsMatchingElement(<Quiz />)).toEqual(true);
  });
  it("Should render a <QuizSetup /> component", () => {
    expect(wrapper.containsMatchingElement(<QuizSetup />)).toEqual(true);
  });
  it("Should render a <NavBar /> component", () => {
    expect(wrapper.containsMatchingElement(<NavBar />)).toEqual(true);
  });
});

describe("Mounted App", () => {
  it("Should render quiz component on route '/quiz'", () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={["/quiz"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.containsMatchingElement(<Quiz />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<QuizSetup />)).toEqual(false);
  });
  it("Should render quizSetup component on route '/'", () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.containsMatchingElement(<Quiz />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<QuizSetup />)).toEqual(true);
  });
  it("Should render a <NavBar /> component on any route", () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.containsMatchingElement(<NavBar />)).toEqual(true);
    wrapper = mount(
      <MemoryRouter initialEntries={["/quiz"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.containsMatchingElement(<NavBar />)).toEqual(true);
  });
  it("Navbar should send user to the homepage when clicked", () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={["/quiz"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.containsMatchingElement(<Quiz />)).toEqual(true);
    expect(wrapper.containsMatchingElement(<QuizSetup />)).toEqual(false);
    wrapper
      .find(NavBar)
      .find(Link)
      .simulate("click", { button: 0 });
    expect(wrapper.containsMatchingElement(<Quiz />)).toEqual(false);
    expect(wrapper.containsMatchingElement(<QuizSetup />)).toEqual(true);
  });

  it("set state function should update app.js state", async () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={["/quiz"]}>
        <App />
      </MemoryRouter>
    );
    let appWrapper = wrapper.find(App).at(0); // Isolates app component
    // Creating test characteristics object for use in setQuizState function:
    const exampleTestCharacteristics = {
      difficulty: "Hard",
      category: "9",
      numOfQuestions: "10",
      numOfPlayers: "2"
    };
    // objective: setquizstate function can take 2 arguments: e.g. difficulty, and difficultyValue. The function should then update quizState object, returning the same object
    // except for the relevant part (e.g. difficulty), which should be updated.
    appWrapper
      .instance()
      .setQuizState("difficulty", exampleTestCharacteristics.difficulty);
    await waitUntil(() => appWrapper.instance().state.quizSetup.difficulty); //async to wait for the api to return!
    expect(appWrapper.instance().state.quizSetup.difficulty).toEqual("Hard");

    appWrapper
      .instance()
      .setQuizState("category", exampleTestCharacteristics.category);
    await waitUntil(() => appWrapper.instance().state.quizSetup.category); //async to wait for the api to return!
    expect(appWrapper.instance().state.quizSetup.category).toEqual("9");

    appWrapper
      .instance()
      .setQuizState(
        "numOfQuestions",
        exampleTestCharacteristics.numOfQuestions
      );
    await waitUntil(() => appWrapper.instance().state.quizSetup.numOfQuestions); //async to wait for the api to return!
    expect(appWrapper.instance().state.quizSetup.numOfQuestions).toEqual("10");

    //currently no dropdown for changing number of players. This test will be relevant when we build on MPV.
    appWrapper
      .instance()
      .setQuizState("numOfPlayers", exampleTestCharacteristics.numOfPlayers);
    await waitUntil(() => appWrapper.instance().state.quizSetup.numOfPlayers); //async to wait for the api to return!
    expect(appWrapper.instance().state.quizSetup.numOfPlayers).toEqual("2");
  });

  it("should pass default quiz setup data to quiz", () => {
    //Default quiz setup data
    const quizInfo = {
      difficulty: "easy",
      category: "9",
      numOfQuestions: "5"
    };

    let wrapper = mount(
      <MemoryRouter initialEntries={["/quiz"]}>
        <App />
      </MemoryRouter>
    );
    let appWrapper = wrapper.find(App).at(0);
    expect(appWrapper.find(Quiz).prop("quizInfo")).toEqual(quizInfo);
  });

  // it("should pass updated quiz setup data from quizSetup to state", () => {
  //   //Updated quiz setup data
  //   const quizInfo = {
  //     difficulty: "hard",
  //     category: "28",
  //     numOfQuestions: "10"
  //   };

  //   let wrapper = mount(
  //     <MemoryRouter initialEntries={["/"]}>
  //       <App />
  //     </MemoryRouter>
  //   );
  //   let appWrapper = wrapper.find(App).at(0);
  //   appWrapper
  //     .find(QuizSetup)
  //     .find(".difficultyDropDown")
  //     .simulate("change", { target: { value: "hard" } });
  //simluate drop down changes to quizInfo values above
  //change to quiz route and test the props of it

  // });
});
