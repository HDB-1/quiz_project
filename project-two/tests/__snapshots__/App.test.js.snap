// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow App Should render correctly 1`] = `
<Component
  className="App"
>
  <NavBar />
  <Switch>
    <Route
      exact={true}
      path="/"
    >
      <QuizSetup
        handleChange={[Function]}
      />
    </Route>
    <Route
      path="/quiz"
    >
      <Quiz
        quizInfo={
          Object {
            "category": "9",
            "difficulty": "easy",
            "numOfQuestions": "5",
          }
        }
      />
    </Route>
    <Route>
      <NotFound />
    </Route>
  </Switch>
</Component>
`;
